spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRowLevelOperationsWithLineage.java

```diff
@@ -0,0 +1,490 @@
+      recordsByPartition.put(spec.specId(), partition, recordsForPartition);
+    }
+
+    return recordsByPartition;
+  }
+
+  protected void appendUnpartitionedRecords(Table table, List<Record> records) throws IOException {
+    appendRecords(table, partitionRecords(records, table.spec(), record -> null));
+  }
+
+  // Append unpartitioned records?
+  protected void appendRecords(Table table, PartitionMap<List<Record>> partitionedRecords)
+      throws IOException {
+    AppendFiles append = table.newAppend();
+
+    for (Map.Entry<Pair<Integer, StructLike>, List<Record>> entry : partitionedRecords.entrySet()) {
+      OutputFile file = Files.localOutput(temp.resolve(UUID.randomUUID().toString()).toFile());
+      DataWriter<Record> writer =
+          new GenericAppenderFactory(schemaWithRowLineage(table.schema()), table.spec())
+              .newDataWriter(
+                  EncryptionUtil.plainAsEncryptedOutput(file), fileFormat, entry.getKey().second());
+      List<Record> recordsForPartition = entry.getValue();
+      writer.write(recordsForPartition);
+      writer.close();
+      append =
+          append
+              .appendFile(writer.toDataFile())
+              .toBranch(branch != null ? branch : SnapshotRef.MAIN_BRANCH);
+    }
+
+    append.commit();
+  }
+
+  protected static Record createRecord(
+      Schema schema, int id, String data, long rowId, long lastUpdatedSequenceNumber) {
+    Record record = GenericRecord.create(schema);
+    record.set(0, id);
+    record.set(1, data);
+    record.set(2, rowId);
+    record.set(3, lastUpdatedSequenceNumber);
+    return record;
+  }
+
+  private Snapshot latestSnapshot(Table table) {
+    return branch != null ? table.snapshot(branch) : table.currentSnapshot();
+  }
+
+  // Expected should have all data columns and the last updated sequence at the end
+  // Actual should all data columns, followed by row ID, last updated sequence, file, and then pos
+  private void assertAddedRowLineage(Object[] expected, Object[] actual) {
+    // validate all the data columns
+    for (int pos = 0; pos < expected.length - 1; pos++) {
+      assertThat(actual[pos]).isEqualTo(expected[pos]);
+    }
+
+    // Determine the expected row ID from the firstRowId of the file and the pos
+    int actualRowIdPos = actual.length - 4;
+    int filePos = actual.length - 2;
+    int rowPositionPos = actual.length - 1;
+    String file = actual[filePos].toString();
+    long fileFirstRowId =
+        (Long)
+            sql("SELECT first_row_id FROM %s.files WHERE file_path = '%s'", selectTarget(), file)
+                .get(0)[0];
+    long rowPosition = (long) actual[rowPositionPos];
+    long expectedFirstRowId = fileFirstRowId + rowPosition;
+    assertThat(actual[actualRowIdPos]).isEqualTo(expectedFirstRowId);
+
+    int expectedLastSequencePos = expected.length - 1;
+    int actualLastSequencePos = 3;
+    assertThat(actual[actualLastSequencePos]).isEqualTo(expected[expectedLastSequencePos]);
```

Comment:
Why not check this in the assertion on line 470? Doesn't it just require selecting `_last_updated_sequence_number` after the content? And, isn't that where this is anyway because `actualLastSequencePos` is hard-coded to 3?